erpnext/accounts/report/accounts_receivable/accounts_receivable.py

```diff
@@ -110,17 +107,22 @@ def get_data(self):
 		# Get Exchange Rate Revaluations
 		self.get_exchange_rate_revaluations()
 
+		self.prepare_ple_query()
 		self.data = []
+		self.voucher_balance = OrderedDict()
 
-		for ple in self.ple_entries:
-			self.update_voucher_balance(ple)
+		with frappe.db.unbuffered_cursor():
```

Comment:
I'd also suggest rewriting the whole logic in SQL. SQL can quite expressive, especially with procedures.

The reason is simple, serializing and transmitting all this data over the wire is still QUITE expensive and slow, if MariaDB executes same logic, it's also doing it at likely 50x-100x faster speeds.

Reply:
@ankush 

I did some quick and dirty raw sql refactor locally and tested. There are some major challenges.
1. Current report makes heavy use of Python Dictionary to calculate allocations for each vouchers.
2. SQL doesn't have an Dictionary data structure. We'll have to use DB Table to emulate that behavior.

As it is complete change in data structure, it requires more time to refactor. So, for now I've laid the ground work by making the data fetching method configurable.

![Screenshot from 2025-05-05 14-00-21](https://github.com/user-attachments/assets/8becf744-6615-498a-b6f2-eb7c481e072b)

Serves 2 purposes,
1. Fail-safe
2. Can be used to compare the output easily and bring it to parity

Reply:
https://github.com/frappe/erpnext/pull/47892